traits/trait_types.py

```diff
@@ -2020,25 +2037,24 @@ def create_editor(self):
             name=self.name,
             cols=self.cols or 3,
             evaluate=self.evaluate,
-            mode=self.mode or "radio",
+            format_func=self.format_func,
+            mode=self.mode if self.mode else "radio",
         )
 
     def _get(self, object, name, trait):
         """ Returns the current value of a dynamic enum trait.
         """
         value = self.get_value(object, name, trait)
-        values = eval(self.values)
+        values = xgetattr(object, self.name)
         if value not in values:
-            value = None
-            if len(values) > 0:
-                value = values[0]
+            value = collection_default(values)
```

Comment:
I was trying to tackle #389 and this line is all I needed to solve #389.

Reply:
@kitchoi can you clarify: if this PR gets merged, does that mean #389 can be closed? Or is there more work to do between merging this PR and solving #389?

Reply:
I think #389 can be closed by this PR (as of now) without more being done to it. 

While #389 concerns the dictionary only, the idea is to relax the requirement on the `values` supporting `__index__`.  As long as `values` implements `__iter__`, Enum can obtain the allowed values. 

I understand your concern with the potential non-deterministic behaviour, e.g. for `set` and the `dict` prior to Python 3.6.  But if, say, `values` is an instance of `OrderedDict`, the `Enum` on master will fail trying to use the first key as a default.  One has to make another property traits to return a list from the keys of `OrderedDict`.  I guess it is about how opinionated this `Enum` should be when it comes to setting a default value.

Reply:
To verify:
```
from traits.api import HasTraits, Enum, Any


class Foo(HasTraits):

    a = Enum(values="b")

    b = Any()

    def _b_default(self):
        return {"a": 1, "b": 2}


if __name__ == "__main__":
    f = Foo()
    # f.configure_traits()
    print(repr(f.a))
```

On this branch, a value `'a'` was printed.
On master, this results in an error:
```
Traceback (most recent call last):
  File "dict_enum.py", line 17, in <module>
    print(repr(f.a))
  File "/Users/kchoi/ETS/traits/traits/trait_types.py", line 2030, in _get
    value = values[0]
KeyError: 0
```

Reply:
Hmm. I think I liked the error better. :-)